test/unit/StakingPool/depositTo.js

```diff
@@ -0,0 +1,425 @@
+    await stakingPool.connect(user).depositTo([
+      {
+        amount,
+        trancheId: firstActiveTrancheId,
+        tokenId: depositNftId,
+        destination,
+      },
+    ]);
+
+    const updatedDepositData = await stakingPool.deposits(depositNftId, firstActiveTrancheId);
+    const newRewardShares = await getNewRewardShares({
+      stakingPool,
+      initialStakeShares: firstDepositData.stakeShares,
+      stakeSharesIncrease: newStakeShares,
+      initialTrancheId: firstActiveTrancheId,
+      newTrancheId: firstActiveTrancheId,
+    });
+    expect(updatedDepositData.pendingRewards).to.equal(0);
+    expect(updatedDepositData.lastAccNxmPerRewardShare).to.equal(0);
+    expect(updatedDepositData.stakeShares).to.equal(firstDepositData.stakeShares.add(newStakeShares));
+    expect(updatedDepositData.rewardsShares).to.equal(firstDepositData.rewardsShares.add(newRewardShares));
+  });
+
+  it('updates deposit stakeShares, rewardShares, pendingRewards and lastAccNxmPerRewardShare', async function () {
+    const { stakingPool } = this;
+    const {
+      members: [user],
+    } = this.accounts;
+
+    const { amount, tokenId, destination, depositNftId } = depositToFixture;
+
+    const { firstActiveTrancheId } = await getTranches();
+
+    // first deposit
+    await stakingPool.connect(user).depositTo([
+      {
+        amount,
+        trancheId: firstActiveTrancheId,
+        tokenId,
+        destination,
+      },
+    ]);
+
+    // Generate rewards
+    const coverRequest = {
+      coverId: 0,
+      productId: 0,
+      amount: parseEther('1'),
+      period: 3600 * 24 * 30, // 30 days
+      gracePeriod: 3600 * 24 * 30,
+      globalCapacityRatio: 20000,
+      capacityReductionRatio: 0,
+      rewardRatio: 5000,
+    };
+    await stakingPool.connect(this.coverSigner).allocateStake(coverRequest);
+
+    await increaseTime(
+      25 * // days
+        24 * // hours
+        60 * // minutes
+        60, // seconds
+    );
+
+    const firstDepositData = await stakingPool.deposits(depositNftId, firstActiveTrancheId);
+    const newStakeShares = await estimateStakeShares({ amount, stakingPool });
+
+    // deposit to the same tokenId
+    await stakingPool.connect(user).depositTo([
+      {
+        amount,
+        trancheId: firstActiveTrancheId,
+        tokenId: depositNftId,
+        destination,
+      },
+    ]);
+
+    const updatedDepositData = await stakingPool.deposits(depositNftId, firstActiveTrancheId);
+    const newRewardShares = await getNewRewardShares({
+      stakingPool,
+      initialStakeShares: firstDepositData.stakeShares,
+      stakeSharesIncrease: newStakeShares,
+      initialTrancheId: firstActiveTrancheId,
+      newTrancheId: firstActiveTrancheId,
+    });
+
+    // TODO: How can I make pendingRewards and lastAccNxmPerRewardShare to not be 0?
+    expect(updatedDepositData.pendingRewards).to.equal(0);
+    expect(updatedDepositData.lastAccNxmPerRewardShare).to.equal(0);
```

Comment:
@shark0der @roxdanila What are the easiest/ideal steps to accumulate rewards so these two variables are set to a value different from 0?

Reply:
The only way to have rewards would be by allocating some capacity (ie cover buy)

Reply:
see the integration tests for buy cover <> claims for a setup example: https://github.com/NexusMutual/smart-contracts/pull/260

Reply:
Thanks! I tried that in this test by calling `allocateStake()` (impersonating the cover contract) between two deposits but wasn't able to get values on those variables different from `0`. Probably I'm doing something wrong, I'll review the setup of the test and try different values and calls

Reply:
cc @danoctavian

Reply:
I wasn't able to make the following variables not be 0.
```
deposit.pendingRewards
deposit.lastAccNxmPerRewardShare
```
After reviewing and debugging, it looks like the root cause is that the global variable `rewardPerSecond` is never set, which is used to calculate `accNxmPerRewardsShare` causing that variable to also be always 0.

The only time the `rewardPerSecond` variable is given a value is on `updateTranches` when expiring a bucket, which reduces its amount: `_rewardPerSecond -= rewardBuckets[_firstActiveBucketId].rewardPerSecondCut;`. 

I think it could be missing the code to increase `rewardPerSecond` with calculated rewards per second in `allocateStake()`.

Also, because of the previous thing mentioned, the two skipped tests right now revert with `Error: VM Exception while processing transaction: reverted with panic code 0x11 (Arithmetic operation underflowed or overflowed outside of an unchecked block)` in the line
`_rewardPerSecond -= rewardBuckets[_firstActiveBucketId].rewardPerSecondCut;`

Adding `rewardPerSecond += _rewardPerSecond;` in allocateStake() makes the test pass, but I'm not sure this is the correct way to fix this.

Reply:
@shark0der  i checked the code as well and it seems that yes we don't have a logic that increments `rewardPerSecond` value as @patitonar says.

Reply:
@patitonar good catch!

1. Yes, the `rewardPerSecond += _rewardPerSecond;` is the correct way to fix - feel free to uncomment that code.

2.  I'm looking at the added `rewards * 10000`. I logged the premium and the rewards and I got 41 and 20 (wei) which is definitely the wrong amount for the requested `1e18` amount. Here's all I logged:
```
console.log("allocated for cover: %s", allocatedCoverAmount);
console.log("initial cap used   : %s", initialCapacityUsed);
console.log("total cap used     : %s", totalCapacity);
console.log("premium            : %s", premium);
console.log("rewards            : %s", rewards);
console.log("_rewardPerSecond   : %s", _rewardPerSecond);

Printed:
allocated for cover: 100
initial cap used   : 0
total cap used     : 200000000
premium            : 41
rewards            : 20
_rewardPerSecond   : 0
```
The bug is likely in the capacity calculations, we'll have to debug that.

3. I think there's one more item that we might want to fix. Given that we divide the total amount of rewards to be distributed to the number of seconds it's being streamed - we'll end up with some precision loss. I guess the biggest error will happen with the smallest cover over the longest period of time, which would be a cover of 0.01 nxm (smallest allocation unit) for 1 year. The cheapest price there can be would be 1% per year, which gets us to a 0.0001 (=1e14) nxm premium. `1e14 - 1e14 // (365 * 24 * 3600) * (365 * 24 * 3600) = 6256000` wei-nxm not being distributed.  That's a 0.000006% error, which is not tragic but we need to address this. I think the proper way to address this is to recalculate the `rewards` amount:
```
uint rewardStreamPeriod = expireAtBucket * BUCKET_DURATION - block.timestamp;
uint _rewardPerSecond = rewards / rewardStreamPeriod;
rewards = _rewardPerSecond * rewardStreamPeriod;
````

@patitonar would you please apply (1) and (3) in the solidity code?

Reply:
Thanks! I'll apply (1) and (3) and dig more on (2) to try to find the bug

Reply:
Applied (1) here https://github.com/NexusMutual/smart-contracts/pull/495/commits/fbec993e9cd01458eeb144408a740b94ae2a4064
Applied (3) here https://github.com/NexusMutual/smart-contracts/pull/495/commits/e373490f618e5979d2173ac4ed309553d814d009

Regarding (2) I reviewed the code, and I'm not 100% but I think I found the bug, although I may be missing something or interpreting something incorrectly. I started following why the `premium` variable was so small.

It is calculated here and it is supposed to be in 18 decimals https://github.com/NexusMutual/smart-contracts/blob/896dbdcb08cc4a89191ef133087d1e9d72ac6517/contracts/modules/staking/StakingPool.sol#L656-L663

The returned value from `getPremium` is based on the result from `calculatePremiumPerYear()` https://github.com/NexusMutual/smart-contracts/blob/896dbdcb08cc4a89191ef133087d1e9d72ac6517/contracts/modules/staking/StakingPool.sol#L1529-L1538

On the tests, it was always returning the `basePremium` variable which is calculated as following https://github.com/NexusMutual/smart-contracts/blob/896dbdcb08cc4a89191ef133087d1e9d72ac6517/contracts/modules/staking/StakingPool.sol#L1541-L1551

Here `basePrice` was expected to be in 18 decimals but instead `basePrice` was always `product.targetPrice` which is a percentage expressed with 2 decimals. So I think the formula is not properly using the `basePrice` variable.

So I applied the following fix https://github.com/NexusMutual/smart-contracts/pull/495/commits/88cf5c269af0d405f5976651f790989436728ee7

```solidity
// scale coverAmount to 18 decimals and apply price percentage
uint basePremium = (coverAmount * NXM_PER_ALLOCATION_UNIT) * basePrice / TARGET_PRICE_DENOMINATOR;
```

After this change, the results were more reasonable and the tests passed. We will need to run some tests with specific numbers to validate that formula is correct when unit testing `allocateStake()` .

Thoughts on this?

Reply:
4. Also, when I was reviewing the code I found that `PRICE_CHANGE_PER_DAY` is 18 decimals. The calculated `priceDrop` is huge compared to `product.nextPrice` and `product.targetPrice` so on the test, it always uses `product.targetPrice` as the `basePrice`. Shouldn't `PRICE_CHANGE_PER_DAY` be expressed or scaled into a similar value like nextPrice and targetPrice? (I guess this also affects `PRICE_BUMP_RATIO`) 

https://github.com/NexusMutual/smart-contracts/blob/896dbdcb08cc4a89191ef133087d1e9d72ac6517/contracts/modules/staking/StakingPool.sol#L1508-L1519

Reply:
Hey @patitonar, very good observations. _I think_ we ended up with these values because we initially had 18 decimals for price (initial price, next price, base price) but changed the variable sizes in structs. Your both suggestions make sense:
- https://github.com/NexusMutual/smart-contracts/commit/88cf5c269af0d405f5976651f790989436728ee7 looks correct to me
- `PRICE_CHANGE_PER_DAY` and `PRICE_BUMP_RATIO` are indeed off. I'll leave a comment with a fix suggestion in a comment below

Reply:
```
diff --git a/contracts/modules/staking/StakingPool.sol b/contracts/modules/staking/StakingPool.sol
index 65492804..16931a84 100644
--- a/contracts/modules/staking/StakingPool.sol
+++ b/contracts/modules/staking/StakingPool.sol
@@ -118,13 +118,13 @@ contract StakingPool is IStakingPool, ERC721 {
   uint public constant INITIAL_PRICE_DENOMINATOR = 100_00;
   uint public constant TARGET_PRICE_DENOMINATOR = 100_00;
 
-  // base price bump is +0.2% for each 1% of capacity used, ie +20% for 100%
-  // 20% = 0.2
-  uint public constant PRICE_BUMP_RATIO = 0.2 ether;
+  // base price bump
+  // +0.2% for each 1% of capacity used, ie +20% for 100%
+  uint public constant PRICE_BUMP_RATIO = 20_00; // 20%
 
   // next price smoothing
-  // 0.005 ether = 0.5% out of 1e18
-  uint public constant PRICE_CHANGE_PER_DAY = 0.005 ether;
+  // 0.5% per day
+  uint public constant PRICE_CHANGE_PER_DAY = 50; // 0.5%
 
   // +2% for every 1%, ie +200% for 100%
   uint public constant SURGE_PRICE_RATIO = 2 ether;
```

Reply:
You can run `cat | git apply`, then paste the above code, followed by a `ctrl+d`

Reply:
Thanks! Applied the fixes for `PRICE_CHANGE_PER_DAY` and `PRICE_BUMP_RATIO` here https://github.com/NexusMutual/smart-contracts/pull/495/commits/4e15371418ff847b4c30d1c0d525bbe7b70091b7